### ISSUE TITLE
=type syntax shadows all the time

### ISSUE THREAD

**Issue by philipcmonk** [urbit/urbit#1352]
`=type` is super convenient, but about 20% of the time the shadowing is super frustrating.  This is obviously bad design.  The old solution to this was to give a unique canonical name for each type (eg `pax` for `path`), but that requires twice the memorization for no benefit.

Many languages keep types and variables in separate namespaces.  This lets you quit having to worry so much about names because there's a canonical answer to "what should a variable of type `t` be called?".  One common way to do this is say that all types start with a capital letter.  Then a variable of type `State` is canonically called `state`.  This is very reasonable, and solves the problem elegantly.

It's against the hoon aesthetic to use capital letters, so I propose suffixing `-` to the type name.  Thus, a variable of type `state` becomes `state-`.  We don't ever use suffixed `-`, so this is easily read as "the variable of type `state`.  I propose using this both at define time and use time.  This solves the problem of "I see variable `a-point` but can't find the definition because it's `=a=point`.

This doesn't syntactically collide with the old way, so we could add this without removing the prefix `=` and see which wins out in practice.

This turns:

```
++  new-event
  |=  [=ship =udiff:point]
  ^+  this-su
  =/  =point  (~(gut by pos.zim.pki) ship *point)
  =/  =diff   (udiff-to-diff:^point udiff point-)
  (public-keys:feel %diff ship diff)
```

into:

```
++  new-event
  |=  [ship- udiff:point-]
  ^+  this-su
  =/  point-  (~(gut by pos.zim.pki) ship- *point)
  =/  diff-   (udiff-to-diff:point udiff- point-)
  (public-keys:feel %diff ship- diff-)
```

I don't actually know how to implement the postfix parsing without running into memory issues.  Prefix is much easier, and would be more in line with the rest of our syntax.  I don't care, as long as we stop shadowing.

**Comment by ohAitch** [comment]
This puts a finger pretty well on my unease with shadow syntax!

For parsing, we probably just want to make a @tas parser that accepts the
trailing - and filters it out in this case(/ rejects otherwise): surely we
want udiff-:point and not udiff:point-

**Comment by ohAitch** [comment]
(To be specific: rejects it in arm names, accepts it in references, converts to foo-=foo in tile syntax)

My one concern would be two-level usage like `mold-`, but that seems easy enough to avoid via renaming / not using the abbreviated face form, if that pops up

**Comment by baudtack** [comment]
Can... you explain why shadowing is annoying? I'm clearly not as experienced but I don't understand the problem.

**Comment by philipcmonk** [comment]
@baudtack 

Consider a function that takes a `point` and modifies it, producing a new one.  Can you spot the bug:

```
|=  [=point =attribute]
^-  point
=.  attribute.point  attribute
point
```

How about if it's async:

```
|=  [=point =attribute]
=/  m  (async:stdio ,point)
^-  form:m
;<  =attribute  bind:m  (validate-attribute attribute)
;<  =point     bind:m  (apply-change point attribute)
(pure:m point)
```

Hint: the first requires one additional `^` and the second requires like three, but in practice it's not worth the `^`'s, so we usually mangle the names.

**Comment by philipcmonk** [comment]
Here's a fun example in code I'm reviewing right now: https://github.com/urbit/urbit/blob/bb69a4b80b75ded7e745559821cbddbc1771e09c/pkg/arvo/app/peerlist.hoon#L160

(Also a few lines later `^tag`.)  How long does it take you to determine which `groups` definition this line refers to?

**Comment by baudtack** [comment]
I mean i think the bug in the first one is pretty obvious, but I was supposed to be looking for it so... The second one is certainly more annoying. I think these are just two examples that should probably not shadow. It's not clear to me how implementing a compiler change would do anything to actually alleviate this that simply not using shadowing when it's inappropriate wouldn't.

**Comment by ohAitch** [comment]
I think an important part here is things being dubiously entwined: if
shadowing was appropriate, does it suddenly become inappropriate when you
add a ^- to the gate? "Gate with a product type matching one of its
arguments" is not exactly uncommon

**Comment by philipcmonk** [comment]
The whole point of the prefix `=` syntax is to not have to name things that don't need a different name.  These are great examples of times when it shouldn't need a new name, but we usually have to choose one anyway.  Shadowing is bad, and you should not habitually need `^`.  This is an anti-feature because it encourages both of those as a "solution" to a real problem.

To put it another way, prefix `=` is a nice feature, but it's not orthogonal to the rest of the language -- it interacts very poorly with other features.  If it didn't shadow, it *would* be orthogonal to rest of the language.

**Comment by baudtack** [comment]
> does it suddenly become inappropriate when you add a ^- to the gate? "Gate with a product type matching one of its arguments" is not exactly uncommon

I mean I'd say that's what ^+ is for but...

> To put it another way, prefix = is a nice feature, but it's not orthogonal to the rest of the language -- it interacts very poorly with other features. If it didn't shadow, it would be orthogonal to rest of the language.

Fair enough. If I understand you want it to instead make the face be sometype- ? Not a fan of the particular thing chosen but it's sort of immaterial.

**Comment by philipcmonk** [comment]
Yeah, I don't care for the `-` suffix aesthetically, but I think this solution is correct up to alpha equivalence.  Maybe use prefix `a-`, so that a variable of type `point` is `a-point` and of type `groups` is `a-groups`.  I could probably get used to that convention.

**Comment by ohAitch** [comment]
I would definitely feel more weird hardcoding english grammar into
hoon.hoon than a glyph, and I think most uses are better characterized as
the- than a- exactly; it's cute though!

Re: prefix, now that $foo has been reclaimed from old tile form, thoughts
on using that? (Bonus points for how =| $foo would be $:foo)

**Comment by philipcmonk** [comment]
`the-` is more accurate, but it's long and "wordy".  `a-` is shorter and less natural, making it more "glyphy".  A more balanced option would be circumfix `-`, as in `-point-`.

A loud glyph suggests this is more than a convention, and that all non-type variables need a `$`.  Plus I'm not sure I can get over the php thing.  Did you know that you can just [add more `$`'s](https://www.php.net/manual/en/language.variables.variable.php)?

**Comment by baudtack** [comment]
Please dear god not $. I'm having php ptsd just thinking about it.

**Comment by philipcmonk** [comment]
> Still, perhaps there's other ways in which we can solve for confusing ^s without introducing a new naming syntax. I'd like to see that explored first before deciding what kind of solution we actually want here.

Identical names for different things is very wrong.  A small, algorithmic perturbation (like captializing the first letter, adding an affix, etc) is good.  I'm generally opposed to adding new possible characters to names, that should be unnecessary.  For the moment, I'm going to start using `=a=point` a lot and see how it feels.

**Comment by loganallenc** [comment]
I personally would much prefer types to be capitalized.
